src/EditorFeatures/Core/EditorConfigSettings/Updater/SettingsUpdateHelper.cs

```
@@ -348,5 +423,93 @@ static bool IsBothVisualBasicAndCSharp(Language language, string[] splicedFileEx
+            sb.Append($"dotnet_diagnostic.{orderedIds.Last()}.severity={severity}");
+            return sb.ToString();
+        }
+
+        private static (string untrimmedKey, string key, string value, string severitySuffixInValue, string commentValue) GetOptionParts(string? curLineText)
+        {
+            var groups = s_optionEntryPattern.Match(curLineText).Groups;
+            var untrimmedKey = groups[1].Value.ToString();
+            var key = untrimmedKey.Trim();
+            var value = groups[2].Value.ToString();
+            var severitySuffixInValue = groups[3].Value.ToString();
+            var commentValue = groups[4].Value.ToString();
+            return (untrimmedKey, key, value, severitySuffixInValue, commentValue);
+        }
+
+        private static bool IsValidHeader(TextLine? mostRecentHeader, TextLine? lastValidHeader)
+        {
+            return mostRecentHeader is not null &&
+                   lastValidHeader is not null &&
+                   mostRecentHeader.Equals(lastValidHeader);
+        }
+
+        private static bool ShouldSetAsLastValidHeader(string curLineText, out string mostRecentHeaderText)
+        {
+            var groups = s_headerPattern.Match(curLineText.Trim()).Groups;
+            mostRecentHeaderText = groups[1].Value.ToString().ToLowerInvariant();
+            return mostRecentHeaderText.Equals("*", StringComparison.Ordinal);
+        }
+
+        private static (string fileName, string[] splicedFileExtensions) GetHeaderParts(string mostRecentHeaderText)
+        {
+            // We splice on the last occurrence of '.' to account for filenames containing periods.
+            var nameExtensionSplitIndex = mostRecentHeaderText.LastIndexOf('.');
+            var fileName = mostRecentHeaderText.Substring(0, nameExtensionSplitIndex);
+            var splicedFileExtensions = mostRecentHeaderText[(nameExtensionSplitIndex + 1)..].Split(',', ' ', '{', '}');
```

What about `[*]`. I expect it will throw `ArgumentOutOfRangeException` since the length passed to Substring is negative.

---

yep adding as part of https://github.com/dotnet/roslyn/issues/52696

---

I've added a test for this case

---

@jmarolf I'm bit confused how the test passes, `nameExtensionSplitIndex` will be -1 since there is no "." in the header, and hence Substring should throw an exception for a negative length. Did the new test come across this code path with `mostRecentHeaderText` equals `[*]`?